// @sap/cli-core/utils/commands.js
import { Command, Option } from "commander";
import { uniq } from "lodash-es";
import { CHARACTERS, } from "../types.js";
import { create as createRootHandler } from "../commands/handler/root/index.js";
import { get } from "../logger/index.js";
const registeredOptions = {};
const getLogger = () => get("utils.commands");
const checkEmpty = (name, value) => {
    if (value.trim() === "") {
        throw new Error(`missing value for parameter ${name}`);
    }
};
const checkCommand = (command) => {
    const { command: name, description } = command;
    try {
        checkEmpty("command", name);
        checkEmpty("description", description);
    }
    catch (err) {
        const { trace } = getLogger();
        trace(`failed to check command ${JSON.stringify(command, null, 2)} because of`, err);
        throw err;
    }
};
export const createCommand = (name) => {
    const { output } = getLogger();
    const command = new Command(name);
    command.enablePositionalOptions();
    command.exitOverride();
    command.configureOutput({ writeErr: output, writeOut: output });
    command.allowUnknownOption();
    command.allowExcessArguments();
    return command;
};
export function getAbsoluteCommandName(command) {
    if (!command.parent) {
        return command.name();
    }
    return `${getAbsoluteCommandName(command.parent)}.${command.name()}`;
}
export const getShortFlagForLongName = (command, option) => {
    const commandName = getAbsoluteCommandName(command);
    for (const s of Object.keys(registeredOptions[commandName] ?? {})) {
        if (registeredOptions[commandName][s] === option.longName) {
            return s;
        }
    }
    throw new Error(`no short flag registered for command ${commandName}, long name ${option.longName}`);
};
export const isOptionAlreadyRegistered = ({ longName }, command) => {
    const opts = command.options?.map((o) => o.long?.replace("--", ""));
    return !!opts?.includes(longName);
};
function shortFlagFromAnyCommand(optionLongName) {
    let shortFlag;
    Object.keys(registeredOptions).find((c) => {
        shortFlag = Object.entries(registeredOptions[c]).find(([_command, longName]) => optionLongName === longName)?.[0];
        return !!shortFlag;
    });
    return shortFlag;
}
function removeUsedShortFlags(possibleShortFlags) {
    const usedShortFlags = uniq(Object.keys(registeredOptions).reduce((p1, c1) => {
        return p1.concat(Object.keys(registeredOptions[c1]));
    }, []));
    return possibleShortFlags.filter((s) => !usedShortFlags.includes(s));
}
const getInitialShortFlagOptions = (option) => {
    return [
        option.longName[0],
        option.longName[0].toUpperCase() === option.longName[0]
            ? option.longName[0].toLowerCase()
            : option.longName[0].toUpperCase(),
        ...CHARACTERS,
    ];
};
export const registerLongName = (command, option) => {
    try {
        return getShortFlagForLongName(command, option);
    }
    catch {
        let possibleShortFlags = getInitialShortFlagOptions(option);
        possibleShortFlags = removeUsedShortFlags(possibleShortFlags);
        if (possibleShortFlags.length === 0) {
            // no short flag is left (possibleShortFlags.length === 0),
            // fall back to the initial set of options
            possibleShortFlags = getInitialShortFlagOptions(option);
        }
        const alreadyUsedShortFlag = shortFlagFromAnyCommand(option.longName);
        if (alreadyUsedShortFlag) {
            possibleShortFlags.unshift(alreadyUsedShortFlag);
        }
        const commandName = getAbsoluteCommandName(command);
        for (const c of possibleShortFlags) {
            if (!registeredOptions[commandName]?.[c]) {
                registeredOptions[commandName] = registeredOptions[commandName] ?? {};
                registeredOptions[commandName][c] = option.longName;
                return c;
            }
        }
        throw new Error(`no short flag found for name ${option.longName}, command ${commandName}`);
    }
};
export const buildOption = async (command, option) => {
    const mandatory = !!option.required;
    const shortFlag = registerLongName(command, option);
    let newDescr = option.description;
    if (!mandatory) {
        newDescr = newDescr ? `${newDescr} (optional)` : "(optional)";
    }
    const argStr = option.args
        ? option.args.reduce((p, c) => (c.optional ? `${p} [${c.name}]` : `${p} <${c.name}>`), "")
        : "";
    const cOption = new Option(`-${shortFlag}, --${option.longName}${argStr}`, newDescr)
        .makeOptionMandatory(mandatory)
        .hideHelp(!!option.hidden);
    if (option.default) {
        const defaultStr = typeof option.default === "function"
            ? await option.default()
            : option.default;
        cOption.default(defaultStr);
    }
    let choicesArr = [];
    if (option.choices) {
        choicesArr =
            typeof option.choices === "function"
                ? await option.choices()
                : option.choices;
    }
    if (choicesArr.length > 0) {
        cOption.choices(choicesArr);
    }
    return cOption;
};
export const buildCommand = async (topCommand, command) => {
    []
        .concat([command], command.type === "topCommand" ? command.subCommands : [])
        .forEach((c) => checkCommand(c));
    const { command: name, description, options } = command;
    const comm = createCommand(name);
    topCommand.addCommand(comm);
    comm.description(description);
    comm.enablePositionalOptions();
    for (const option of options ?? []) {
        // eslint-disable-next-line no-await-in-loop
        comm.addOption(await buildOption(comm, option));
    }
    if (command.type === "topCommand") {
        const { subCommands } = command;
        for (const subCommand of subCommands) {
            // eslint-disable-next-line no-await-in-loop
            await buildCommand(comm, subCommand);
        }
    }
    else {
        // command.type === "command"
        const { args } = command;
        if (args) {
            args.forEach((a) => comm.argument(`<${a.argument}>`, a.description));
        }
        const rootHandler = await createRootHandler(command, `Failed to ${command.description}`)(comm);
        comm.action(rootHandler);
    }
    return comm;
};
