// @sap/cli-core/utils/http/index.js
import axios from "axios";
import https from "https";
import { HttpsProxyAgent } from "https-proxy-agent";
import { getPackageName, getVersion } from "../../config/core.js";
import { get as getConfig, set as setConfig } from "../../config/index.js";
import { DEFAULT_TLS_VERSION, X_CSRF_TOKEN } from "../../constants.js";
import { get } from "../../logger/index.js";
import { logVerbose } from "../../logger/utils.js";
import { createHttpsAgent, getTlsVersion, isLegacyTlsDetectionEnabled, isTlsProtocolError, } from "./httpsAgent.js";
import { printCorrelationId, printError } from "./utils.js";
export const DEFAULTS = {
    maxBodyLength: -1,
    maxContentLength: -1,
    maxRedirects: 0, // must be set to 0 because maxBodyLength: -1 won't work otherwise, see github.com/axios/axios/issues/4263,
};
const HEADERS_ETAG = "x-sap-cli-core-discovery-etag";
const getLogger = () => get("http");
const setEtag = (headers) => {
    const { debug } = getLogger();
    if (headers?.[HEADERS_ETAG]) {
        setConfig({ etag: headers[HEADERS_ETAG] });
    }
    else {
        debug("header %s is not available", HEADERS_ETAG);
    }
};
const getCsrfTokenFromConfig = () => {
    const cnfg = getConfig();
    return cnfg[X_CSRF_TOKEN] ? { [X_CSRF_TOKEN]: cnfg[X_CSRF_TOKEN] } : {};
};
const createHttpsAgentWithTlsVersion = (proxy, tlsVersion) => {
    const agentOptions = {
        minVersion: tlsVersion,
        maxVersion: tlsVersion,
        rejectUnauthorized: true,
    };
    if (proxy) {
        return new HttpsProxyAgent(proxy, agentOptions);
    }
    return new https.Agent(agentOptions);
};
const logProxyConfiguration = (proxy) => {
    const logger = getLogger();
    if (proxy) {
        logVerbose(logger, "using https proxy agent for https proxy");
        logger.debug(`using https proxy agent for https proxy ${proxy}`);
    }
    else {
        logger.debug("no https proxy defined via environment variable https_proxy");
    }
};
const logTlsConfiguration = () => {
    const logger = getLogger();
    const tlsVersion = getTlsVersion();
    const useLegacyTls = isLegacyTlsDetectionEnabled();
    logger.debug(`TLS configuration - version: ${tlsVersion}, legacy detection: ${useLegacyTls}`);
};
const buildAxiosConfig = (config, httpsAgent) => {
    return {
        httpsAgent,
        proxy: false,
        ...config,
        ...DEFAULTS,
        headers: {
            ...getCsrfTokenFromConfig(),
            ...config.headers,
            "User-Agent": `${getPackageName()}+${getVersion()}`,
        },
    };
};
const logSuccessResponse = (response) => {
    const cnfg = getConfig();
    const logger = getLogger();
    if (cnfg.verbose) {
        logVerbose(logger, "%d %s", response.status, response.statusText);
        printCorrelationId(response);
    }
    logger.trace("response", response);
};
const logErrorResponse = (err) => {
    const cnfg = getConfig();
    const logger = getLogger();
    logger.error("error while executing http request", err.toString(), err.response);
    if (cnfg.verbose) {
        printError(err);
        printCorrelationId(err.response);
    }
};
const shouldAttemptTlsFallback = (error) => {
    const tlsVersion = getTlsVersion();
    const isUsingDefaultTls = tlsVersion === DEFAULT_TLS_VERSION;
    const isTlsError = isTlsProtocolError(error);
    return isUsingDefaultTls && isTlsError && !isLegacyTlsDetectionEnabled();
};
const executeRequest = async (config, httpsAgent) => {
    const axiosConfig = buildAxiosConfig(config, httpsAgent);
    return axios(axiosConfig);
};
const attemptTlsFallback = async (config, originalError) => {
    const logger = getLogger();
    const { debug, warn, error } = logger;
    warn("TLSv1.3 connection failed with protocol error, automatically retrying with TLSv1.2");
    debug("original error:", originalError?.message ??
        JSON.stringify(originalError));
    try {
        const HTTPS_PROXY = process.env.https_proxy ?? process.env.HTTPS_PROXY;
        const fallbackAgent = createHttpsAgentWithTlsVersion(HTTPS_PROXY, "TLSv1.2");
        debug("retrying request with TLSv1.2");
        const res = await executeRequest(config, fallbackAgent);
        const cnfg = getConfig();
        if (cnfg.verbose) {
            logVerbose(logger, "%d %s (fallback to TLSv1.2)", res.status, res.statusText);
            printCorrelationId(res);
        }
        debug("TLSv1.2 fallback successful");
        logger.trace("response", res);
        return res;
    }
    catch (fallbackErr) {
        error("TLSv1.2 fallback also failed", fallbackErr.toString(), fallbackErr.response);
        throw fallbackErr;
    }
};
export const fetch = async (config) => {
    const logger = getLogger();
    try {
        logVerbose(logger, "%s %s", config.method.toUpperCase(), config.url);
        logger.debug("http config: %s", JSON.stringify(config));
        const HTTPS_PROXY = process.env.https_proxy ?? process.env.HTTPS_PROXY;
        const httpsAgent = createHttpsAgent(HTTPS_PROXY);
        logProxyConfiguration(HTTPS_PROXY);
        logTlsConfiguration();
        const res = await executeRequest(config, httpsAgent);
        logSuccessResponse(res);
        setEtag(res.headers);
        return res;
    }
    catch (err) {
        logErrorResponse(err);
        setEtag(err?.response?.headers);
        if (shouldAttemptTlsFallback(err)) {
            const fallbackResponse = await attemptTlsFallback(config, err);
            setEtag(fallbackResponse.headers);
            return fallbackResponse;
        }
        throw err;
    }
};
