// @sap/cli-core/index.js
import * as dotenv from "dotenv";
import { configureLoggers } from "./configureLoggers.js";
import { set } from "./config/index.js";
import { AuthenticationMethod, CLI_DEPRECATED, CLI_DEPRECATION_MESSAGE, CLI_DESCRIPTION, CLI_DISCOVERY_PATHS, CLI_GENERIC_OPTIONS_HELP, CLI_NAME, CLI_PACKAGE_NAME, CLI_SAP_HELP, CLI_SUPPORTED_AUTHENTICATION_METHODS, CLI_VERSION, } from "./constants.js";
export { run } from "./dwc/run.js";
export { configure, getCommands } from "./module.js";
export { AuthenticationMethod };
export { getVersion, getBin, getDescription, getName, getPackageName, } from "./utils/utils.js";
dotenv.config({
    quiet: true, // default is false, but we want to avoid logging to the stdout
});
configureLoggers();
export const configureCoreModule = ({ name, packageName, discoveryPaths, description, sapHelpLink, authenticationMethods = [AuthenticationMethod.oauth], version, genericOptionsHelp, deprecated = false, deprecationMessage, }) => {
    set({
        [CLI_NAME]: name,
        [CLI_PACKAGE_NAME]: packageName,
        [CLI_DESCRIPTION]: description,
        [CLI_DISCOVERY_PATHS]: discoveryPaths,
        [CLI_SAP_HELP]: sapHelpLink,
        [CLI_SUPPORTED_AUTHENTICATION_METHODS]: authenticationMethods,
        [CLI_VERSION]: version,
        [CLI_DEPRECATED]: deprecated,
        [CLI_DEPRECATION_MESSAGE]: deprecationMessage,
        [CLI_GENERIC_OPTIONS_HELP]: genericOptionsHelp,
    });
};
