// @sap/cli-core/types.js
// unfortunately there does not seem to be a good solution to automatically
// create the below array from a statically defined type. Thus, the array must
// always be kept in sync with the type Secret defined above.
export const ALLOWED_SECRET_TYPES = [
    "id",
    "tenantUrl",
    "customClient",
    "client_id",
    "client_secret",
    "access_token",
    "expires_in",
    "expires_after",
    "refresh_token",
    "token_url",
    "authorization_url",
    "token_type",
    "jti",
    "scope",
    "id_token",
    "authorization_flow",
];
export var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["INACTIVE"] = 1] = "INACTIVE";
    LogLevel[LogLevel["ERROR"] = 2] = "ERROR";
    LogLevel[LogLevel["WARN"] = 3] = "WARN";
    LogLevel[LogLevel["INFO"] = 4] = "INFO";
    LogLevel[LogLevel["DEBUG"] = 5] = "DEBUG";
    LogLevel[LogLevel["TRACE"] = 6] = "TRACE";
})(LogLevel || (LogLevel = {}));
/* jscpd:ignore-start */
export const CHARACTERS = [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g",
    // "h", skip 'h' to avoid confusion with '--help' option
    "i",
    "j",
    "k",
    "l",
    "m",
    "n",
    "o",
    "p",
    "q",
    "r",
    "s",
    "t",
    "u",
    "v",
    "w",
    "x",
    "y",
    "z",
    "A",
    "B",
    "C",
    "D",
    "E",
    "F",
    "G",
    "H",
    "I",
    "J",
    "K",
    "L",
    "M",
    "N",
    "O",
    "P",
    "Q",
    "R",
    "S",
    "T",
    "U",
    "V",
    "W",
    "X",
    "Y",
    "Z",
];
export var GrantType;
(function (GrantType) {
    GrantType["authorization_code"] = "authorization_code";
    GrantType["client_credentials"] = "client_credentials";
    GrantType["refresh_token"] = "refresh_token";
})(GrantType || (GrantType = {}));
